crates/runtime/src/tools/extension/terminal/manager.rs

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+use std::collections::HashMap;
+use std::env;
+use std::io::{self};
+use std::process::Stdio;
+use std::sync::Arc;
+use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
+use tokio::process::{Child, Command};
+use tokio::sync::Mutex;
+
+pub struct TerminalManager {
+    terminals: Arc<Mutex<HashMap<usize, Child>>>,
+    next_id: Arc<Mutex<usize>>,
+}
+
+pub static END_OF_COMMAND_MARKER: &str = "\x1e__CMD_END__\x1f";
+
+impl Default for TerminalManager {
+    fn default() -> Self {
+        TerminalManager {
+            terminals: Arc::new(Mutex::new(HashMap::new())),
+            next_id: Arc::new(Mutex::new(1)),
+        }
+    }
+}
+
+impl TerminalManager {
+    fn get_shell_command() -> Command {
+        let default_shell = if cfg!(target_os = "windows") {
+            env::var("COMSPEC").unwrap_or("cmd.exe".to_string())
+        } else {
+            env::var("SHELL").unwrap_or("bash".to_string())
+        };
+        Command::new(default_shell)
+    }
+
+    /// Spawns a new terminal session and adds it to the manager.
+    ///
+    /// # Returns
+    ///
+    /// * `io::Result<usize>` - The unique ID of the new terminal or an I/O error.
+    pub async fn spawn_terminal(&self) -> io::Result<usize> {
+        // Generate a unique ID for the new terminal
+        let mut id_lock = self.next_id.lock().await;
```

Comment:
The alternatively can be AtomicUsize that supports thread safe incremental updates